addons/binding/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/nhc2/NikoHomeControlCommunication2.java

```diff
@@ -0,0 +1,812 @@
+    }
+
+    @Override
+    public synchronized void stopCommunication() {
+        if (communicationStarted != null) {
+            communicationStarted.complete(false);
+        }
+        communicationStarted = null;
+        mqttConnection.stopPublicConnection();
+        mqttConnection.stopProfileConnection();
+    }
+
+    @Override
+    public boolean communicationActive() {
+        CompletableFuture<Boolean> started = communicationStarted;
+        if (started == null) {
+            return false;
+        }
+        try {
+            // Wait until we received all devices info to confirm we are active.
+            return started.get(5000, TimeUnit.MILLISECONDS);
+        } catch (InterruptedException | ExecutionException | TimeoutException e) {
+            logger.debug("Niko Home Control: exception waiting for connection start");
+            return false;
+        }
+    }
+
+    /**
+     * After setting up the communication with the Niko Home Control Connected Controller, send all initialization
+     * messages.
+     *
+     */
+    private void initializePublic() throws MqttException {
+        NhcMessage2 message = new NhcMessage2();
+
+        message.method = "systeminfo.publish";
+        mqttConnection.publicConnectionPublish("public/system/cmd", gson.toJson(message));
+
+        message.method = "profiles.list";
+        mqttConnection.publicConnectionPublish("public/authentication/cmd", gson.toJson(message));
+    }
+
+    /**
+     * After setting up the profile communication with the Niko Home Control Connected Controller, send all profile
+     * specific initialization messages.
+     *
+     */
+    private void initializeProfile() throws MqttException {
+        NhcMessage2 message = new NhcMessage2();
+
+        message.method = "services.list";
+        mqttConnection.profileConnectionPublish(profileUuid + "/authentication/cmd", gson.toJson(message));
+
+        message.method = "devices.list";
+        mqttConnection.profileConnectionPublish(profileUuid + "/control/devices/cmd", gson.toJson(message));
+
+        message.method = "locations.list";
+        mqttConnection.profileConnectionPublish(profileUuid + "/control/locations/cmd", gson.toJson(message));
+
+        message.method = "notifications.list";
+        mqttConnection.profileConnectionPublish(profileUuid + "/notification/cmd", gson.toJson(message));
+    }
+
+    private void connectionLost() {
+        logger.debug("Niko Home Control: connection lost");
+        stopCommunication();
+        handler.controllerOffline();
+    }
+
+    private void timePublishEvt(String response) {
+        Type messageType = new TypeToken<NhcMessage2>() {
+        }.getType();
+        List<NhcTimeInfo2> timeInfo = null;
+        try {
+            NhcMessage2 message = gson.fromJson(response, messageType);
+            if (message.params != null) {
+                timeInfo = message.params.stream().filter(p -> (p.timeInfo != null)).findFirst().get().timeInfo;
```

Comment:
If you use `orElse` instead of `get` you don't have to catch `NoSuchElementException`.

Reply:
@Hilbrand trying this, but I cannot pass null to orElse because of Null annotations. Did I miss something?

Reply:
Ah. ok those null checks are sometimes hard... I did some checking the `orElse` is not going to help here either. So it might be better to put the `get()` output in a variable and check if that is not null. It's a bit cleaner than to rely on catching exceptions. Or just leave the exception, but put a comment in the empty block about why it's catching.

Reply:
Ok, I put in some comments. Checking null is cumbersome as the variable would have to be a variable of a nested class object.